field/TinyLang/Field/Evaluator.hs

```diff
@@ -144,57 +187,75 @@ evalBinOp Mul x y = return . UniConst Field $ x `mul` y
 evalBinOp Div x y = UniConst Field <$> divEval x y
 evalBinOp BAt x y = asIntegerEval x >>= fmap (UniConst Bool) . flip atEval y . fromIntegral
 
-evalStatementUni
-    :: (MonadEvalError f m, Eq f, Field f, AsInteger f)
-    => Env (SomeUniConst f) -> Statement f -> Expr f a -> m (UniConst f a)
-evalStatementUni env (ELet (UniVar _ var) def) rest = do
-    defR <- evalExprUni env def
-    evalExprUni (insertVar var (Some defR) env) rest
-evalStatementUni env (EAssert expr) rest = do
-    exprR <- evalExpr env expr
-    if exprR
-        then evalExprUni env rest
-        else throwError $ AssertionFailedEvalError expr
+evalProgramUni :: (Monad m, Eq f, Field f, AsInteger f)
+    => Program f -> EvalT f m ()
+evalProgramUni = evalStatementsUni . unProgram
+
+evalStatementsUni :: (Monad m, Eq f, Field f, AsInteger f)
+    => Statements f -> EvalT f m ()
+evalStatementsUni stmts
+        | ELet (UniVar _ var) def : rest <- unStatements stmts = do
+            defR <- evalExprUni def
+            -- Record the store in the
+            -- TODO:  Fix a potential O(n^2) overhead
+            modify (++[Some defR])
```

Comment:
I'd just collect the trace in reverse order, i.e.

```haskell
modify (Some defR :)
```

because we only care about traces staying the same after renaming, so their order doesn't matter.

Or you could use `DList`.